File: app-mvvm-fsharp/ViewLocator.fs
// Copyright (c) The Avalonia Project. All rights reserved.
// Licensed under the MIT license. See licence.md file in the project root for full license information.

namespace AvaloniaAppTemplate

open System
open Avalonia.Controls
open Avalonia.Controls.Templates
open AvaloniaAppTemplate.ViewModels

type ViewLocator () =
    interface IDataTemplate with
        member __.SupportsRecycling with get() = false
        member __.Match data = data :? ViewModelBase
        member __.Build data : IControl =
            let name = data.GetType().FullName.Replace("ViewModel", "View")
            let _type = Type.GetType name
            match _type with
            | null -> TextBlock (Text = "Not Found: " + name) :> IControl
            | _    -> (Activator.CreateInstance _type) :?> IControl

File: templates/csharp/xplat/AvaloniaTest/ViewModels/ViewLocator.cs
using System;
using Avalonia.Controls;
using Avalonia.Controls.Templates;

namespace AvaloniaTest.ViewModels;

[StaticViewLocator]
public partial class ViewLocator : IDataTemplate
{
	public Control Build(object? data)
	{
		if (data == null) throw new ArgumentNullException(nameof(data));
		var type = data.GetType();
		if (s_views.TryGetValue(type, out var func))
		{
			return func.Invoke();
		}
		throw new Exception($"Unable to create view for type: {type}");
	}

	public bool Match(object? data)
	{
		return data is ViewModelBase;
	}
}


